include/picongpu/particles/atomicPhysics/kernel/CheckForOverSubscription.kernel

```diff
@@ -118,18 +114,79 @@ namespace picongpu::particles::atomicPhysics::kernel
+            DataSpace<picongpu::simDim> const superCellCellOffset = superCellIdx * picongpu::SuperCellSize::toRT();
+            auto forEachCell = pmacc::lockstep::makeForEach<T_FieldEnergyUseCache::numberCells, T_Worker>(worker);
+            forEachCell(
+                [&worker,
+                 &superCellCellOffset,
+                 &eFieldBox,
+                 &eFieldEnergyUseCache,
+                 &sharedResourcesOverSubscribed,
+                 &rejectionProbabilityCache](uint32_t const linearCellIndex)
+                {
+                    DataSpace<picongpu::simDim> const localCellIndex
+                        = pmacc::math::mapToND(picongpu::SuperCellSize::toRT(), static_cast<int>(linearCellIndex));
+                    DataSpace<picongpu::simDim> const cellIndex = localCellIndex + superCellCellOffset;
+
+                    // sim.unit.charge()^2 * sim.unit.time()^2 / (sim.unit.mass() * sim.unit.length()^3)
+                    //  * sim.unit.length()^3
+                    // = sim.unit.charge()^2 * sim.unit.time()^2 / sim.unit.mass()
+                    constexpr float_X eps0HalfTimesCellVolume
+                        = (picongpu::sim.pic.getEps0() / 2._X) * picongpu::sim.pic.getCellSize().productOfComponents();
+
+                    // sim.unit.charge()^2 * sim.unit.time()^2 / sim.unit.mass()
+                    //  * ((sim.unit.mass() * sim.unit.length())/(sim.unit.time()^2 * sim.unit.charge()))^2
+                    // = sim.unit.charge()^2 * sim.unit.time()^2 * sim.unit.mass()^2 * sim.unit.length()^2
+                    //  / (sim.unit.mass() * sim.unit.time()^4 * sim.unit.charge()^2)
+                    // = sim.unit.mass() * sim.unit.length()^2/ (sim.unit.time()^2 * sim.unit.length())
+                    // sim.unit.energy()
+                    float_X const eFieldEnergy = eps0HalfTimesCellVolume * pmacc::math::l2norm2(eFieldBox(cellIndex));
+
+                    // eV * 1 = eV * sim.unit.energy()/sim.unit.energy() = (ev / sim.unit.energy()) * sim.unit.energy()
+                    // sim.unit.energy()
+                    float_X const eFieldEnergyUse
+                        = picongpu::sim.pic.get_eV() * eFieldEnergyUseCache.energyUsed(linearCellIndex);
+
+                    float_X rejectionProbability;
+
+                    if(eFieldEnergyUse > eFieldEnergy)
+                    {
+                        // cell is oversubscribed by suggested changes
+
+                        // calculate rejection probability
+                        rejectionProbability = pmacc::math::max(
+                            // proportion of weight we want to reject
+                            (eFieldEnergyUse - eFieldEnergy) / eFieldEnergyUse,
+                            // but approximately at least one average one macro ion per cell should be rejected
+                            1._X / static_cast<float_X>(sim.getTypicalNumParticlesPerCell()));
```

Comment:
Can be `sim.getTypicalNumParticlesPerCell() < 1`? Then you would always have a rejection probability `> 1`...

Reply:
Good Catch, as far as I know we do not enforce that `sim.getTypicalNumParticlesPerCell() >= 1`, but we probably should enforce it here at compile time.

Reply:
fix per `PMACC_CASSSERT(.....);`